### PR TITLE
CollectionFS  cfs:tempstore and cfs:worker to fix concurrency issues with GridFS uploads

### PR DESCRIPTION
**Pull request by lcampanis**
This PR fixes concurrency issues with CollectionFS and GridFS adapter in Meteor and Reaction Commerce.

We have brought in cfs:tempstore and cfs:worker in `./packages` and modified them to fix the following errors:

**cfs:tempstore**
```
-----------FINISH STREAM _tempstore
---SA REMOVE
/opt/reaction/dist/bundle/programs/server/node_modules/fibers/future.js:280
                        throw(ex);
                        ^

TypeError: Cannot read property 'keys' of undefined
    at packages/cfs_tempstore.js:326:63
    at runWithEnvironment (packages/meteor.js:1176:24)
```

**cfs:worker** (see issue https://github.com/reactioncommerce/reaction/issues/2071)
In this fix we just wrap all DB calls in try/catch. This ensures tempstore doesn't crash the container app when this exception takes place. Although this fix doesn't exactly fix the errors, it does ensure that any upload actually completes, since the workers will always make sure all uploads continue streaming as if nothing happened. We have tested heavily a max of 5 concurrent image uploads from 1MB to 15Mb with an external mongoDB. The final result is "so far" positive. 
```
/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/utils.js:98 process.nextTick(function() { throw err; });
MongoError: no chunks found for file, possibly corrupt at Function.MongoError.create (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/error.js:31:11)
/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:825:29
/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/gridfs/grid_store.js:1201:5 at handleCallback (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/utils.js:96:12)
/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/collection.js:1401:5
at handleCallback (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/utils.js:96:12)
at /app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb/lib/cursor.js:675:5
at handleCallback (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/cursor.js:172:5)
at setCursorNotified (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/cursor.js:517:3)
at /app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/cursor.js:590:16
at queryCallback (/app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/cursor.js:233:18)
at /app/bundle/programs/server/npm/node_modules/meteor/cfs_gridfs/node_modules/mongodb-core/lib/connection/pool.js:436:18
at nextTickCallbackWith0Args (node.js:489:9)
at process._tickDomainCallback (node.js:459:13)
```

Both fixes are interlinked and without one, it will produce an error for the other one. Together it seems all problems are fixed. 

Credit @artlimes 

PS. We have also implemented `gridfs-locking-stream` with GridFS and will share a new repo as soon as we've tested everything in more depth.

### CODE REVIEW

**Comment by spencern** [comment]
I'm still getting seeing this error on the `cfs-fix` branch
```
/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/utils.js:98
    process.nextTick(function() { throw err; });
                                  ^
MongoError: no chunks found for file, possibly corrupt
    at Function.MongoError.create (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/error.js:31:11)
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/gridfs/grid_store.js:825:29
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/gridfs/grid_store.js:1201:5
    at handleCallback (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/utils.js:96:12)
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/collection.js:1401:5
    at handleCallback (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/utils.js:96:12)
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb/lib/cursor.js:675:5
    at handleCallback (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/cursor.js:172:5)
    at setCursorNotified (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/cursor.js:517:3)
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/cursor.js:590:16
    at queryCallback (/Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/cursor.js:233:18)
    at /Users/sn/dev/reaction/.meteor/local/isopacks/cfs_gridfs/npm/node_modules/mongodb-core/lib/connection/pool.js:436:18
    at nextTickCallbackWith0Args (node.js:489:9)
    at process._tickCallback (node.js:418:13)
=> Exited with code: 1
```

The server is killed and does restart eventually, but as far as I can tell, the images are not persisted.

Reproduction steps:
1. Create external MongoDB server.
2. Run Reaction using that external MongoDB server.
3. Duplicate the default product.
4. Using 3 separate chrome windows, as close to simultaneously as possible, upload 3 images (1 each window). Uploading two images to Product, and one image to Product Copy.

**Comment by spencern** [comment]
I'm also wondering if it's possible that this has anything to do with the amount of Memory in the MongoDB instance.

Here's a screenshot from the compose instance I'm using to test

image

Of note, the MEMORY FAILCNT which spiked pretty hard.

File: packages/worker/fileWorker.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  });
+
+  // Initiate observe for finding files that have been stored so we can delete
+  // any temp files
+  fsCollection.files.find(getDoneQuery(fsCollection.options.stores)).observe({
+    added: function(fsFile) {
+      FS.debug && console.log("FileWorker ADDED - calling deleteChunks for", fsFile._id);
+      try {
+        FS.TempStore.removeFile(fsFile);
+      }catch(err) {
```
Comment (spencern) [inline packages/worker/fileWorker.js]: Reaction coding style prefers to have spaces after the brackets } and before the parens
e.g.
`} catch (err) {`

File: packages/worker/fileWorker.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  FS.debug && console.log('saving to store ' + storeName);
+
+  try {
+
+    var writeStream = storage.adapter.createWriteStream(fsFile);
+    var readStream = FS.TempStore.createReadStream(fsFile);
+
+    // Pipe the temp data into the storage adapter
+    readStream.pipe(writeStream);
+  }catch(err){
```
Comment (spencern) [inline packages/worker/fileWorker.js]: Reaction style prefers spaces after brackets and parens
e.g `} catch (err) {`

**Comment by lcampanis** [comment]
@spencern linted now.

**Comment by aaronjudd** [comment]
@spencern @lcampanis I guess we should decide to merge this or not...  cons -> bring all this extra code in, and I think @spencern is still able to replicate the issue with this update. Pro: @lcampanis says it fixed the issue for them...

**Comment by spencern** [comment]
I don't think it hurts to merge it. I don't remember any regressions and it's unlikely that we will miss a CFS update (since there are none) by forking this. 

Since it solved problems for @lcampanis then it's possible my issues were related but different and it actually does solve some problems.

I have started working on the CFS replacement which hopefully will make all of this moot in the near future.

**Comment by lcampanis** [comment]
@aaronjudd @spencern I can confirm that so far we've had no issues with cfs uploads going up to 10Mb from users over slow networks. A cfs replacement is more than necessary as it affects performance dramatically. 
We will be moving cfs into its own DB today, most probably by introducing something like `MONGO_FILES_URL=mongoFileUrl`, unless @jshimko has already started on something similar?

**Comment by zenweasel** [comment]
@aaronjudd @spencern Was there a reason we were still waiting on this?

**Comment by spencern** [comment]
@zenweasel - We can merge.

**Pull request merged**